### DISCUSSION TITLE
Increase of default max number of pages

### DISCUSSION THREAD

**Discussion by bqzero** [HASwitchPlate/openHASP#770] [Polls]
Since there are more and more devices comming that do feature more powerfull configurations (strong CPU, more Flash, hugh PSRAM,...) like for example the Elecrow 7" panel it would be good to extend the default max number of pages (currently set to 12).

That device for example can handle much more than 12 pages and I for example am using it as a main controll on every floor in my house. Letting me control shutters, lights, get sensor informations etc. since I am limited to 12 pages I either have to compile it myself or an increase of this parameter can be done in the default config.
Since there is not much of an issue by increasing that number (no bigger firmware for example) why not doing so.

**Comment by fvanroie** [comment]
Did you actually this this? A jump from 12 pages to 50 or 100 seems rather steep...
Anyways, there definitely are resource implications and [code restrictions to need to be adjusted](https://github.com/HASwitchPlate/openHASP/issues/407) too.

**Reply by bqzero** [reply]
I see, did not find the data type assigned to the page and to answer your question I have not yet adjusted it. I do agree with you that this would in fact have a bigger impact. Never the less the limitation to 12 pages is very low. A 5 bit or 6 bit integer would use more space, but the amount used should be worth it. and as the case you shared the link shows that the devices could handle it (right away not with actions) but they are capable of that.
How about an optional switch that sets the definition by default 4bit and if necessary increase it to 6bit? that way it could be controled (maybe even by board/ressource definition) and assure proper functionality?

also pages as far as I can find are define as I can see as usigned 8bit integers, so only actions is the issue and not matching this
![image](https://github.com/user-attachments/assets/49a2b3bb-912f-4d34-90b5-bef46e31efbc)

also the documentations mentions pages are only limited by available ressources. This somehow contradicts that if only 4bit actions are possible.

**Reply by bqzero** [reply]
Is there a way to place a button on the screen that basically calls for an mqtt publish maybe with some HA configuration. that way we could call what ever page number we need without a change of HASP_NUM_PAGES to workaround that limitation. creating as many pages as you want is possible and calling each with command as well.

**Reply by bqzero** [reply]
BTW as a workaround this is doing the job
define a btn on your page something along the line like this 
![image](https://github.com/user-attachments/assets/56ed4c7b-a13d-4b48-a2b8-708794918d3a)
important add no action to it
and do the configuration in HA like this
![image](https://github.com/user-attachments/assets/f160e5e4-f582-4f9d-a32f-bf7b1d13c9e9)
with payload you can specify the page to be loaded, without the limitation of actions. 

in case 12 is not sufficient do not forget to compile the firmware yourself and use the build flags override for -D HASP_NUM_PAGES=99. I just used 99 to be safe. in fact the firmware size is not changed by that value at all.

**Reply by fvanroie** [reply]
You should be able to define an action locally by using a JSON object: `"action":{"down": "page 32"}`
This should work independent of the `actionid` limitations. But testing is needed...

`actionid` with 4-bit mask is a legacy way to switch pages, but this is now superseded by `action` as a JSON object.
In term, this should be removed from the code.
